lib/Source/AutoHog/AutoHogImporter.cpp

```diff
@@ -0,0 +1,516 @@
+    SmallVector<int> &result) {
+  for (rapidjson::Value::ConstMemberIterator itr = coefficients.MemberBegin();
+       itr != coefficients.MemberEnd(); ++itr) {
+    const char *portName = itr->name.GetString();
+    const int coeffValue = itr->value.GetInt();
+    assert(inputPortNameToOperandIndex.contains(portName) &&
+           "Coefficient port name not found in inputPortNameToOperandIndex");
+    result[inputPortNameToOperandIndex.at(portName)] = coeffValue;
+  }
+}
+
+OwningOpRef<Operation *> translateFromAutoHog(llvm::StringRef inputString,
+                                              MLIRContext *context) {
+  // FIXME: shouldn't have to manually load these dialects!
+
+  context->getOrLoadDialect<arith::ArithDialect>();
+  context->getOrLoadDialect<cggi::CGGIDialect>();
+  context->getOrLoadDialect<func::FuncDialect>();
+  context->getOrLoadDialect<lwe::LWEDialect>();
+  context->getOrLoadDialect<tensor::TensorDialect>();
+
+  LLVM_DEBUG(llvm::dbgs() << "Translating from AutoHoG JSON to MLIR\n");
+  LLVM_DEBUG(llvm::dbgs() << "Input string: \n" << inputString << "\n");
+  const char *json = inputString.data();
+  rapidjson::StringStream ss(json);
+  rapidjson::Document document;
+  document.ParseStream(ss);
+
+  LLVM_DEBUG({
+    rapidjson::StringBuffer buffer;
+    rapidjson::Writer<rapidjson::StringBuffer> writer(buffer);
+    document.Accept(writer);
+    const char *output = buffer.GetString();
+    llvm::dbgs() << "Parsed JSON: " << output << "\n";
+  });
+  assert(document.IsObject() && "JSON failed to parse");
+
+  StringMap<Port> ioPorts;
+  parseIOPorts(document, ioPorts);
+
+  OpBuilder builder(context);
+  ModuleOp moduleOp = builder.create<ModuleOp>(builder.getUnknownLoc());
+  moduleOp->setAttr(
+      "cggi.circuit_name",
+      StringAttr::get(context, document["circuit_name"].GetString()));
+
+  builder.setInsertionPointToStart(&moduleOp.getBodyRegion().front());
+
+  // Parse "ports" field to assemble input and output types
+  // The function will always be a tensor<Nxi1> -> tensor<Mxi1>
+  // for N = #inputs and M = #outputs
+  // The maps created here will be used to access the input and output tensors
+  // during operation creation.
+  StringMap<int> inputPortToTensorIndex;
+  StringMap<int> outputPortToTensorIndex;
+  int numInputs = 0;
+  int numOutputs = 0;
+  for (auto &[portName, port] : ioPorts) {
+    if (port.type == PortType::INPUT) {
+      inputPortToTensorIndex[portName] = numInputs++;
+    } else {
+      outputPortToTensorIndex[portName] = numOutputs++;
+    }
+  }
+  // FIXME: detect proper minBitWidth from the circuit
+  int minBitWidth = 3;
+  Type ciphertextType = lwe::LWECiphertextType::get(
+      context, lwe::UnspecifiedBitFieldEncodingAttr::get(context, minBitWidth),
+      lwe::LWEParamsAttr());
+  Type inputType = RankedTensorType::get({numInputs}, ciphertextType);
+  Type outputType = RankedTensorType::get({numOutputs}, ciphertextType);
```

Comment:
I was just curious why make the choice to use single tensor to hold all the inputs - do we even know how the input bits are distributed in different multi-bit inputs? Or how they should be ordered within the tensor?

I wonder if it might also be worth annotating the function inputs with the port names from the JSON (but that would only be possible if the inputs were unpacked)

Reply:
Maybe I am misunderstanding the autohog work, but I was under the  impression that the inputs to the circuit are still single bit ciphertexts.

Reply:
No! You are probably right, I think what I'm asking for on second thought would be work that we would need to handle if we export a circuit to an AutoHoG input - we probably want to keep track of some circuit metadata so we can map inputs to their single bit ciphertexts.

The question was more about packing them all into a single argument rather than having unpacked single bit ciphertext inputs. The former I guess just makes the IR easier to read? It won't matter if we use that metadata to create client input/output handlers in the end.

Reply:
Yeah, I basically just picked arbitrarily. I was thinking about how we often use `tensor<lwe.ciphertext>`, and I kind of expect these IRs to be quite big (a 64-bit adder, e.g.). We can put whatever metadata we want on the function (a DictionaryAttr mapping int to string, for example). I feel like the right time to do that is when the need arises.

Reply:
Agreed - noted for now it's somewhat arbitrary and when we have a full integration it might make sense to use metadata/reconsider